### PR TITLE
Upgrade Typescript to v5

### PR DESCRIPTION
**Pull request by sisou**
v4 -> v5 was no changes actually, see the commit history.

~This is based on the `usdc` branch, so can only be merged after #465 is merged.~

### CODE REVIEW

File: src/components/RecoveryWords.js [hunk 1]
```diff
@@ -157,9 +165,10 @@ class RecoveryWords extends Nimiq.Observable {
     _setFocusToNextInput(index, paste) {
         index = Math.max(index, 0);
         if (index < this.$fields.length) {
-            this.$fields[index].focus();
-            if (paste) {
-                this.$fields[index].fillValueFrom(paste);
+            const field = this.$fields[index];
+            field.focus();
+            if (paste && 'fillValueFrom' in field) {
```
Comment (onmax) [inline src/components/RecoveryWords.js:170]:
```suggestion
            if (paste && 'fillValueFrom' in field && typeof field.fillValueFrom === 'function') {
```

not sure if this is relevant

File: src/components/RecoveryWords.js [hunk 2]
```diff
@@ -123,18 +130,19 @@ class RecoveryWords extends Nimiq.Observable {
...
-            const mnemonic = this.$fields.map(field => field.value);
+            const mnemonic = this.$fields.map(field => ('value' in field ? field.value : '-'));
             const type = Nimiq.MnemonicUtils.getMnemonicType(mnemonic); // throws on invalid mnemonic
             this._mnemonic = { words: mnemonic, type };
             this.fire(RecoveryWords.Events.COMPLETE, mnemonic, type);
         } catch (e) {
-            if (e.message !== 'Invalid checksum') {
+            const err = /** @type {Error} */ (e);
+            if (err.message !== 'Invalid checksum') {
                 console.error(e); // eslint-disable-line no-console
```
Comment (onmax) [inline src/components/RecoveryWords.js]:
```suggestion
                console.error(err); // eslint-disable-line no-console
```

_Super picky comment_

I wouldn't use `e` after defining `err` with a type

File: tsconfig.json [hunk 3]
```diff
@@ -1,8 +1,8 @@
 {
   "compilerOptions": {
     /* Basic Options */
-    "target": "ES2015",                       /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017','ES2018' or 'ESNEXT'. */
-    "module": "commonjs",                     /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', or 'ESNext'. */
+    "target": "ES2020",                       /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017','ES2018' or 'ESNEXT'. */
```
Comment (onmax) [inline tsconfig.json]:
```suggestion
    "target": "ES2020",                       /* Specify ECMAScript target version. */
```

File: tsconfig.json [hunk 4]
```diff
@@ -1,8 +1,8 @@
...
+    "module": "CommonJS",                     /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', or 'ESNext'. */
```
Comment (onmax) [inline tsconfig.json:6]:
```suggestion
    "module": "CommonJS",                     /* Specify module code generation. */
```

**Review by onmax — APPROVED** [review]
LGTM

File: tsconfig.json (hunk 3 above)
Comment (onmax) [reply tsconfig.json]: People can learn what values are valid from the docs so we don't have these comments out of sync ?

**Review by danimoh — APPROVED** [review]
Cool, thanks for this.
I added my improvements, feel free to have a look.
The change of static properties of classes to actual js `static`s requires an eslint update, so this will be in a separate PR.

@onmax please also review my changes, if you want.

**Comment by danimoh** [comment]
The force-push is for rebase onto master, with resolved conflicts in the `FiatApi`.
However, some additional type errors seem to have surfaced on the latest changes from master.
Taking a look now.

File: src/lib/LoginFile.js [hunk 5]
```diff
@@ -65,9 +65,12 @@ class LoginFile {
...
         await this._drawPromise;
 
-        return new Promise(resolve => {
+        return new Promise((resolve, reject) => {
             this.$canvas.toBlob(blob => {
-                if (!blob) throw new Error('Cannot generate URL');
+                if (!blob) {
+                    reject(new Error('Cannot generate URL'));
+                    return;
+                }
```
Comment (sisou) [inline src/lib/LoginFile.js]: TIL throwing in the `new Promise()` callback doesn't reject the promise. The more you know...

File: tsconfig.json [hunk 6]
```diff
@@ -4,11 +4,6 @@
     "target": "ES2020",                       /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017','ES2018' or 'ESNEXT'. */
     "module": "CommonJS",                     /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', or 'ESNext'. */
     "lib": ["DOM"],                           /* Specify library files to be included in the compilation. */
-    "skipLibCheck": true,                     /* Skip checking .d.ts files. We need this currently, because types in */
-                                              /* lib.webworker.d.ts referenced in ServiceWorker.js clash with types */
-                                              /* in lib.dom.d.ts, and also because our typescript version is too old */
-                                              /* to handle the types of ethers. */
-                                              /* TODO once typescript has been updated, this can be removed. */
```
Comment (sisou) [inline tsconfig.json:11]: Well spotted! :+1:

File: src/common.js [hunk 7]
```diff
@@ -88,8 +88,7 @@ async function runKeyguard(RequestApiClass, opts) { // eslint-disable-line no-un
 
     // Back arrow functionality
     document.body.addEventListener('click', event => {
-        // @ts-ignore (Property 'matches' does not exist on type 'EventTarget'.)
-        if (!event.target || !event.target.matches('a.page-header-back-button')) return;
+        if (!(event.target instanceof HTMLElement && event.target.matches('a.page-header-back-button'))) return;
```
Comment (sisou) [inline src/common.js]: I liked the previous double-negative OR syntax better.

File: src/components/PasswordBox.js [hunk 8]
```diff
@@ -41,7 +41,7 @@ class PasswordBox extends Nimiq.Observable {
 
         this.$el.classList.toggle('hide-input', this.options.hideInput);
 
-        this._passwordInput = new PasswordInput(this.$el.querySelector('[password-input]'));
+        this._passwordInput = new PasswordInput(/** @type {HTMLElement} */(this.$el.querySelector('[password-input]')));
```
Comment (sisou) [inline src/components/PasswordBox.js:44]: Why is this change necessary? I thought with your changes `null | undefined` are also allowed?

File: src/components/PasswordSetterBox.js [hunk 9]
```diff
@@ -27,7 +27,7 @@ class PasswordSetterBox extends Nimiq.Observable {
         this.$el = PasswordSetterBox._createElement($el, this.options);
 
         this._passwordInput = new PasswordInput(
-            this.$el.querySelector('[password-input]'),
+            /** @type {HTMLElement} */ (this.$el.querySelector('[password-input]')),
```
Comment (sisou) [inline src/components/PasswordSetterBox.js:30]: Same here.

File: src/components/PasswordSetterBox.js [hunk 10]
```diff
@@ -52,7 +52,7 @@ class PasswordSetterBox extends Nimiq.Observable {
     }
 
     /**
-     * @param {?HTMLFormElement} $el
+     * @param {?HTMLFormElement | undefined} $el
```
Comment (sisou) [inline src/components/PasswordSetterBox.js:55]: What's the difference between `| undefined` in the type and putting `$el` in square brackets?

File: src/common.js (hunk 7 above)
Comment (danimoh) [reply src/common.js]: Me theoretically too, I only changed it to make it fit into a single line.

File: src/components/PasswordBox.js (hunk 8 above)
Comment (danimoh) [reply src/components/PasswordBox.js:44]: I think that it worked previously, is actually a typescript bug. `querySelector` returns `Element | null`, `PasswordInput` previously accepted `?HTMLElement` which is equivalent to `HTMLElement | null`, which should refuse a passed value of `Element | null` but didn't. Now it's `{?HTMLElement} [$el]`, which is equivalent to `HTMLElement | null | undefined`, which somehow trigger typscript to actually complain (i.e. it's now actually more restrictive), which is why a type assertion is now required. This is the way it should be imo.

File: src/components/PasswordSetterBox.js (hunk 9 above)
Comment (danimoh) [reply src/components/PasswordSetterBox.js:30]: https://github.com/nimiq/keyguard/pull/467#discussion_r1746968333

File: src/components/PasswordSetterBox.js (hunk 10 above)
Comment (danimoh) [reply src/components/PasswordSetterBox.js:55]: `[$el]` is an optional parameter which can be omitted. A parameter with `| undefined` can be set to `undefined` but not omitted. An optional parameter can not precede a non-optional parameter, which is why here it needs to be declared as `| undefined`.

In the old ts version, there was also another difference, which I now didn't observe anymore with the newer version: https://github.com/nimiq/keyguard/pull/461#discussion_r1045729111

File: src/common.js (hunk 7 above)
Comment (sisou) [reply src/common.js]: Changed it back to double negative, that version is only one character longer and still fits into one line.

File: src/components/RecoveryWords.js (hunk 1 above)
Comment (sisou) [reply src/components/RecoveryWords.js:170]: I don't think this is necessary.

File: src/common.js (hunk 7 above)
Comment (danimoh) [reply src/common.js]: Oh, right :+1:
Sometimes in my IDE lines look longer when it shows names of the passed parameters.
